File: ruby/triangle/triangle.rb
class Triangle < Array
  def equilateral?
    uniq.length == 1 if valid?
  end

  def isosceles?
    uniq.length <= 2 if valid?
  end

  def scalene?
    uniq.length == 3 if valid?
  end

private

  def valid?
    return false unless size == 3
    return false unless all?(&:positive?)
    triangular_inequality_holds?
  end

  def triangular_inequality_holds?
    *short_sides, long_side = *sort
    short_sides.sum >= long_side
  end
end

File: haskell/triangle/src/Triangle.hs
module Triangle (TriangleType(..), triangleType) where

import Data.List

data TriangleType = Equilateral
                  | Isosceles
                  | Scalene
                  | Illegal
                  deriving (Eq, Show)

triangleType :: (Num a, Eq a, Ord a) => a -> a -> a -> TriangleType
triangleType 0 0 0 = Illegal
triangleType a b c
  | any (<= 0) sides = Illegal
  | x + y < z = Illegal
  | x == y && y == z = Equilateral
  | x == y || y == z = Isosceles
  | otherwise = Scalene
  where sides@(x:y:z:[]) = sort [a, b, c]
